### ISSUE TITLE
ATProto => ActivityPub @-mentions

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#959]
Not working yet. Example [Bluesky post](https://bsky.app/profile/snarfed.bsky.social/post/3kq6zmsmpc52z) and [bridged fediverse post](https://indieweb.social/@snarfed.bsky.social@atproto.brid.gy/112276964069556117). Guess I'll go back and refresh my memory on #493.

<img width="575" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/6f9a2e83-1e7b-434f-aeb9-eec1a7435311">

---

Bluesky, `at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3kq6zmsmpc52z`:
```json
{
  "$type" : "app.bsky.feed.post",
  "createdAt" : "2024-04-15T19:47:50.404Z",
  "text" : "eh indeed, cc @snarfed.indieweb.social.ap.brid.gy eh"
  "facets" : [{
      "$type" : "app.bsky.richtext.facet",
      "features" : [{
          "$type" : "app.bsky.richtext.facet#mention",
          "did" : "did:plc:bg5udl25mvzg3rt7l5n2hzet"
        }],
      "index" : {
        "byteEnd" : 49,
        "byteStart" : 14
      }
    }
  ],
}
```

AS1:
```json
{
  "objectType": "comment",
  "id": "at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3kq6zmsmpc52z",
  "url": "https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3kq6zmsmpc52z",
  "content": "eh indeed, cc @snarfed.indieweb.social.ap.brid.gy eh",
  "tags": [{
      "objectType": "mention",
      "url": "https://bsky.app/profile/did:plc:bg5udl25mvzg3rt7l5n2hzet",
      "startIndex": 14,
      "displayName": "@snarfed.indieweb.social.ap.brid.gy",
      "length": 35
    }],
}
```

AS2:
```json
{
    "type": "Note",
    "id": "https://atproto.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3kq6zmsmpc52z",
    "url": "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3kq6zmsmpc52z",
    "content": "eh indeed, cc @snarfed.indieweb.social.ap.brid.gy eh",
    "tag": [{
        "startIndex": 14,
        "length": 35,
        "type": "Mention",
        "name": "@snarfed.indieweb.social.ap.brid.gy",
        "href": "https://atproto.brid.gy/ap/https://bsky.app/profile/did:plc:bg5udl25mvzg3rt7l5n2hzet"
      },{
        "type": "Mention",
        "href": "https://indieweb.social/users/snarfed"
      }],
}
```

**Comment by qazmlp** [comment]
I did a bit of sleuthing about how Mastodon parses mentions in `"content"`. In short:

- I don't think it does normally, these really are just links given to the clients.
- Text mentions are [enhanced into links](https://github.com/mastodon/mastodon/blob/0622107449e72d35b22afeeba2f0ba983e914803/app/lib/html_aware_formatter.rb#L35-L37) through [this `extract_mentions_or_lists_with_indices` function](https://github.com/mastodon/mastodon/blob/0622107449e72d35b22afeeba2f0ba983e914803/app/lib/extractor.rb#L10) if they are matched [by this regex](https://github.com/mastodon/mastodon/blob/0622107449e72d35b22afeeba2f0ba983e914803/app/models/account.rb#L70-L71) and [the account is preloaded for HTML formatting](https://github.com/mastodon/mastodon/blob/0622107449e72d35b22afeeba2f0ba983e914803/app/helpers/formatting_helper.rb#L22) (which appears to be the author + to active mentions), but
  - [**This transformation is not applied to remote status' content!**](https://github.com/mastodon/mastodon/blob/main/app/lib/html_aware_formatter.rb#L17-L24)

So in short:

I think you can keep the mention format in the text, but **you'll have to turn it into an HTML link so that it renders properly** towards Mastodon clients and the web interface.

I think the **mentions in the `"tag"` property are most likely fine already**, as they are turned into account references for any further processing [purely by resolving the href](https://github.com/mastodon/mastodon/blob/6b33d3f81be36b772c0f0436c14a7a742062e2de/app/lib/activitypub/activity/create.rb#L224-L233), so that Mastodon should prefill them in `@user@domain`-format when replying to this status.

**Comment by snarfed** [comment]
Thanks!

When we tested and got these working from web a while back in https://github.com/snarfed/bridgy-fed/issues/493#issuecomment-1897843483, the HTML link in content evidently wasn't necessary, but `tag` definitely was.

Also, the goal here is more than just a clickable link; it's to make Mastodon actually handle it as a mention, so that when you click on it, it loads the mentioned user in that local instance's UI, instead of navigating your browser to the remote instance.

(`startIndex` and `length` are BF-specific warts from the AS1 => AS2 conversion, they exist in AS1 but not AS2. OK technically `startIndex` exists for collection pages, but not objects like this. We should be able to ignore them here.)

**Comment by qazmlp** [comment]
Right, `tag` will give you the semantic mention (notifications, reply autofill I think), unless that isn't reliably working either.

I understood this issue here to be specifically about the inline presentation, which is largely handled separately, and for which I'm pretty certain either the link is necessary or *maybe* `@user@domain` formatting of the handle in the text could also work.

**Comment by qazmlp** [comment]
> Also, the goal here is more than just a clickable link; it's to make Mastodon actually handle it as a mention, so that when you click on it, it loads the mentioned user in the local instance's UI, instead of navigating your browser to the remote instance.

I finally found the place this is done, by searching for the `status-mention` class those links have in the web UI. It's matched client-side to the href in `tag`:

https://github.com/mastodon/mastodon/blob/caad1e2628c78d2dfcd86e3a79b0849dc89baf16/app/javascript/mastodon/components/status_content.jsx#L104-L130

This also confirms that the surrounding markup doesn't matter, you just need an `<a>` tag at least for Mastodon. The account @ that's inserted into the new local href is the one Mastodon resolved from the `tag` `href` too, if I didn't completely misread that earlier.

**Comment by snarfed** [comment]
Looked at this a bit. We need to convert the mention `href` to the mentioned actor's AP id. Right now we have eg ATProto record `at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3ksveg5r2qs24`:a

```json
{
    "$type": "app.bsky.feed.post",
    "text": "maybe if @snarfed.indieweb.social.ap.brid.gy and #alf meet up",
    "createdAt": "2024-05-20T03:49:56.091Z",
    "facets": [{
      "$type": "app.bsky.richtext.facet",
      "features": [{
          "$type": "app.bsky.richtext.facet#mention",
          "did": "did:plc:bg5udl25mvzg3rt7l5n2hzet"
        }],
      "index": {
        "byteEnd": 44,
        "byteStart": 9
      }
    }],
  }
```

...to AS2:

```json
{
  "id": "https://bsky.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3ksveg5r2qs24",
  "url": "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3ksveg5r2qs24",
  "content": "<p>maybe if <a href=\"https://bsky.brid.gy/ap/https://bsky.app/profile/did:plc:bg5udl25mvzg3rt7l5n2hzet\">@snarfed.indieweb.social.ap.brid.gy</a> and <a href=\"https://bsky.brid.gy/hashtag/alf\">#alf</a> meet up</p>",
  "published": "2024-05-20T03:49:56.091Z",
  "attributedTo": "https://bsky.brid.gy/ap/did:plc:3ljmtyyjqcjee2kpewgsifvb",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Note",
  "tag": [{
      "type": "Mention",
      "name": "@snarfed.indieweb.social.ap.brid.gy",
      "href": "https://bsky.brid.gy/ap/https://bsky.app/profile/did:plc:bg5udl25mvzg3rt7l5n2hzet"
    }],
}
```

**Comment by mackuba** [comment]
So that should eventually show the mention as "@snarfed@indieweb.social" in the post text on the Mastodon side, right?

**Comment by snarfed** [comment]
Yes, or just `@snarfed`. More importantly, it should _be_ a mention, ie link to that fediverse account inside the local instance, as usual.

**Comment by snarfed** [comment]
^ Progress. We're now at least generating the right `Mention` tag to notify the mentioned user. Example: https://bsky.app/profile/snarfed.bsky.social/post/3ktxejevy3s2o => https://indieweb.social/@snarfed.bsky.social@bsky.brid.gy/112547944255049449

The link doesn't get rewritten to point to the mentioned user inside the local instance, though. It's a normal HTML link that goes to the user on their own instance. Example of how I'd ideally like it to work: https://indieweb.social/@tchambers/112542225021901875 . I've gotten this working before with Web => AP, not sure what needs to change. Does the mention link text need to be the AP handle? Something else? Back to #493, my old nemesis.

**Comment by qazmlp** [comment]
I'm pretty sure this is because you're using `"name":"@snarfed.indieweb.social.ap.brid.gy",` instead of `"name":"@snarfed@indieweb.social",` for the `"Mention"`-`"tag"`.

Remember that these get matched `href=`-`"href":`, **not** text-to-`"name"`!

**Comment by snarfed** [comment]
> text-to-`"name"` as in the standard!

@qazmlp out of curiosity, are you talking about this section? https://www.w3.org/TR/activitystreams-vocabulary/#microsyntaxes

>  A typical social software implementation would typically render such a content such that "@sally" is replaced with a hyperlink to "@sally"'s social profile page and "#givingthanks" is replaced with a hyperlink to a listing of other notes that have been "tagged" with the same topic. Most implementations would also send a special notification to sally letting her know that a note mentioning her has been created.

Not normative, but still, useful if so, thanks for the nudge.

**Comment by qazmlp** [comment]
I'm talking about "EXAMPLE 158" which appears to have the post content as plaintext in JSON:

> ```json
> {
>   "@context": "https://www.w3.org/ns/activitystreams",
>   "name": "A thank-you note",
>   "type": "Note",
>   "content": "Thank you @sally for all your hard work! #givingthanks",
>   "tag": [
>     {
>       "type": "Mention",
>       "href": "http://example.org/people/sally",
>       "name": "@sally"
>     },
>     {
>       "id": "http://example.org/tags/givingthanks",
>       "name": "#givingthanks"
>     }
>   ]
> }
> ```

I'm not sure if any behaviour is actually being prescribed there, though, and in my opinion automatic parsing of incoming plaintext without markup isn't *usually* a great idea, generally speaking.

**Comment by snarfed** [comment]
> I'm not sure if any behaviour is actually being prescribed there, though, and in my opinion automatic parsing of incoming plaintext without markup isn't _usually_ a great idea, generally speaking.

Right! That whole section is non-normative, so no, they're not prescribing behavior per se, and they agree with you on avoiding parsing text content. From the beginning of the section:

> Many social software systems use special text-based microsyntaxes that allow users to define special addressing for notifications, linking, or categorization within objects. For example, including text such as "@username" within an object's content will often route the object to a special "mentions" or "inbox" stream for a particular user. Likewise, including text such as " #topic" within the object's content will often mark the object as being related to the topic "topic"...
>
> While such microsyntaxes MAY be used within the values of the [content](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-content), [name](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-name), and [summary](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-summary) properties on an Activity Streams [Object](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-object), implementations SHOULD NOT be required to parse the values of those properties in order to determine the appropriate routing of notifications, categorization or linking between objects. Instead, publishers SHOULD make appropriate use of the vocabulary terms provided specifically for these purposes.

**Comment by snarfed** [comment]
Looking at this again. Here's the (pruned) AS2 object for the example above with the mention that notified the Mastodon user but didn't get rewritten. The mention `href` is the mentioned user's AP id, and the `href` and `name` both match the link `href` and text in the HTML content. 🤷

```json
{
   "@context" : "https://www.w3.org/ns/activitystreams",
   "type" : "Note",
   "id" : "https://bsky.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3ktxejevy3s2o",
   "url" : "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3ktxejevy3s2o"
   "attributedTo" : "https://bsky.brid.gy/ap/did:plc:3ljmtyyjqcjee2kpewgsifvb",
   "content" : "<p><a href=\"https://indieweb.social/users/snarfed\">@snarfed.indieweb.social.ap.brid.gy</a> and <a href=\"https://bsky.brid.gy/hashtag/alf\">#alf</a> should meet up again</p>",
   "tag" : [{
         "href" : "https://indieweb.social/users/snarfed",
         "name" : "@snarfed.indieweb.social.ap.brid.gy",
         "type" : "Mention"
      },{
         "href" : "https://bsky.brid.gy/hashtag/alf",
         "name" : "#alf",
         "type" : "Hashtag"
      }],
}
```

**Comment by snarfed** [comment]
I think this is now the same as #887.

**Comment by snarfed** [comment]
Closing, will track the local rewriting in #887.
